design-proposals/2022-11-29-metal-port.md

```diff
@@ -0,0 +1,485 @@
+
+The goal here is to move as many setup stages as possible off the main thread.  That should include all the buffer allocation and setup and some of the indirect command buffer encoding.  Direct encoding is going to still occur on the main thread.
+
+#### Required Changes
+
+It's not clear what changes will need to be made here.  The process goes like this:
+- Find a good test case.  Usually something that involves a lot of zooming and panning.
+- Run the test case while recording one or more of: memory or CPU
+- Look for a bottleneck
+- Fix the bottleneck
+
+Typically bottlenecks are where the main CPU is waiting on something it shouldn't be, or tiles are only being loaded on a single thread or other similar issues.  It's never obvious until you start staring at the traces.
+
+#### Benefits
+
+MapLibre Native makes some use of threads, but not as much as it should.  With OpenGL it's not using shared contexts, though it does a lot of work in the Layers to prep for the GL work that has to be done on the main thread.  So it's not terrible, but it obviously could be better.
+
+Mobile devices, even the really cheap ones, now have a lot of CPU cores and a lot of opportunity for work to be done off the main thread.  Maps have an embarassing amount of high level parallelism.  It's worth periodically analyzing the performance to see what might be updated for data loading.
+
+The benefits are faster loading, lower latency, better responsiveness to the user.
+
+### Shader optimization
+
+_Addresses Core Functionality [#7](#core)._
+
+The first version of shaders will work with the direct rendering in Metal.  They'll be simple manual translations from the GLSL in the existing toolkit and they'll look and act similarly to the old shaders.
+
+The second version of the shaders will consist of changes to support indirect rendering.  The way buffers are passed in is different when doing indirect rendering in a way that's not obvious.
+
+Further, this is the time to break the shaders to modular pieces to be reused by developers outside the toolkit.  Passing in a new shader to replace existing functionality works a lot better when you can call sub-routines to implement that MapLibre Native basics for you.
+
+#### Required Changes
+
+This pass is the last version of the shaders and would consist of:
+- Modularization.  We can make it easier to reuse shaders by sharing code where useful.  Even putting the core of a given shader into a subroutine would make it easier to reuse.
+- Data driven styling.  How much of that logic can be pushed into the shaders?  The more work they do, the less the CPU does.
+- Optimization.  Apple provides tools for analyzing time spent in the shaders.  Some optimization here is always useful.
+- Packaging.  Normally Metal shaders are pre-compiled and shipped with the app binary.  We may want to support a pure source version.
```

Comment:
Given past experience with Metal shaders, I have to wonder what all else they include in the compiled blob as I've only seen precompiled Metal shaders be larger on disk compared to source code. We should find out exactly what the size of precompiled Metal shaders would look like before we decide if that is something we want to support exclusively.

Interesting tidbit: [Apple](https://developer.apple.com/documentation/metal/shader_libraries/minimizing_the_binary_size_of_a_shader_library?changes=_3&language=objc) even seems to advise compiling on demand to reduce app size. How fun.
> For some apps, it may be more practical to compile a shader on the device when the app is running, typically to reduce the app’s storage size.

Reply:
I haven't looked closely, but it may be support for multiple GPU generations.

Well, there's no reason we couldn't stick the shader source into a text file and compress it.  If that's how it has to work, then so be it.  But we could at least try to make sure debugging works seamlessly.

I'm also not sure how we could share functions outside the library if a developer is writing their own shaders.  The trick would be to try each of those use cases.